cranelift/codegen/src/egraph.rs

```diff
@@ -1,342 +1,437 @@
-                                .egraph
-                                .add(
-                                    Node::Result {
-                                        value: id,
-                                        result: i,
-                                        ty,
-                                    },
-                                    &mut self.node_ctx,
-                                )
-                                .get();
-                            self.stats.node_created += 1;
-                            self.stats.node_result += 1;
-                            trace!("build: value {} -> id {}", result, projection);
-                            value_to_id.insert(result, projection);
+                if is_pure_for_egraph(cursor.func, inst) {
+                    // Insert into GVN map and optimize any new nodes
+                    // inserted (recursively performing this work for
+                    // any nodes the optimization rules produce).
+                    let mut ctx = OptimizeCtx {
+                        func: cursor.func,
+                        value_to_opt_value: &mut value_to_opt_value,
+                        gvn_map: &mut gvn_map,
+                        eclasses: &mut self.eclasses,
+                        rewrite_depth: 0,
+                        subsume_values: FxHashSet::default(),
+                        remat_values: &mut self.remat_values,
+                        stats: &mut self.stats,
+                    };
+                    let inst = NewOrExistingInst::Existing(inst);
+                    ctx.insert_pure_enode(inst);
+                    // We've now rewritten all uses, or will when we
+                    // see them, and the instruction exists as a pure
+                    // enode in the eclass, so we can remove it.
+                    cursor.remove_inst_and_step_back();
+                } else {
+                    self.stats.skeleton_inst += 1;
+
+                    // Not pure, but may still be a load or store:
+                    // process it to see if we can optimize it.
+                    if let Some(new_result) = self.alias_analysis.process_inst(
+                        cursor.func,
+                        &mut alias_analysis_state,
+                        inst,
+                    ) {
+                        self.stats.alias_analysis_removed += 1;
+                        let result = cursor.func.dfg.inst_results(inst)[0];
+                        value_to_opt_value[result] = new_result;
+                        cursor.remove_inst_and_step_back();
+                    } else {
+                        // Set all results to identity-map to themselves
+                        // in the value-to-opt-value map.
+                        for &result in cursor.func.dfg.inst_results(inst) {
+                            value_to_opt_value[result] = result;
+                            self.eclasses.add(result);
```

Comment:
Could we wrap this up into `process_skeleton_inst` or something like that, just to make the overall algorithm a little easier to read at a glance?